display_list/geometry/dl_path.h

```diff
@@ -35,12 +37,21 @@ class DlPath {
   DlRect GetBounds() const;
 
   bool operator==(const DlPath& other) const;
+  bool operator!=(const DlPath& other) const { return !(*this == other); }
 
   bool IsConverted() const;
+  bool IsVolatile() const;
 
  private:
-  const SkPath sk_path_;
-  mutable impeller::Path path_;
+  struct Data {
+    explicit Data(const SkPath& path) : sk_path(path) {}
+
+    SkPath sk_path;
+    impeller::Path path;
+    uint32_t use_count = 0u;
+  };
+
+  std::shared_ptr<Data> data_;
```

Comment:
Why is this a shared_ptr if the CanvasPath class stores sk path and impeller_path directly?

Reply:
We copy these objects around all over. Everything that needs to be shared between them needs to be encapsulated here in some "shareable" form.

CP stores SkPath directly to edit it. It is copied (invoking copy-on-write for the original SkPath) into the DlPath object when it is retrieved. That DlPath object then is responsible for marking its own copy of SkPath as non-volatile and for creating the iPath as needed.

Until you modify the CanvasPath again, all future retrievers will get a copy of the same DlPath that was first handed out and so they will share the use count and the converted path.